### PR TITLE
Add device memory

### PR DESCRIPTION
**Pull request by jc9123**
Modified fetch_azure.py so that the catalog now contains device memory for GPU. 

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/clouds/service_catalog/data_fetchers/fetch_gcp.py [hunk 1]
```diff
@@ -513,6 +513,23 @@ def get_catalog_df(region_prefix: str) -> pd.DataFrame:
...
+        'L4': 24,
+        'A100': 40,
+        'A100-80GB': 80,
+        'A100-40GB': 40,
+        'T4': 16,
+        'P4': 8,
+        'V100': 16,
+        'P100': 16,
+        'K80': 12,
+        '': ''
```
Comment (infwinston) [inline sky/clouds/service_catalog/data_fetchers/fetch_gcp.py:526]: sorry why this `''`?

File: sky/clouds/service_catalog/data_fetchers/fetch_gcp.py [hunk 2]
```diff
@@ -513,6 +513,23 @@ def get_catalog_df(region_prefix: str) -> pd.DataFrame:
     # Round the prices.
     df['Price'] = df['Price'].round(PRICE_ROUNDING)
     df['SpotPrice'] = df['SpotPrice'].round(PRICE_ROUNDING)
+    gpu_map = {
```
Comment (infwinston) [inline sky/clouds/service_catalog/data_fetchers/fetch_gcp.py:516]: Could you add a reference link with comment above?

File: sky/clouds/service_catalog/data_fetchers/fetch_azure.py [hunk 3]
```diff
@@ -244,11 +244,79 @@ def get_additional_columns(row):
         axis='columns',
     )
 
+    def create_gpu_map(df):
+        # Map of Azure's machine with GPU to their corresponding memory
+        # Result is hard-coded since Azure's API to not return such info
+        # may be outdated so need to be maintained
```
Comment (infwinston) [inline sky/clouds/service_catalog/data_fetchers/fetch_azure.py:250]: Could you add a reference link on how these information are found? also, how did we make sure we cover all the instance types?

File: sky/clouds/service_catalog/data_fetchers/fetch_azure.py [hunk 4]
```diff
@@ -244,11 +244,79 @@ def get_additional_columns(row):
...
+            'Standard_NC64as_T4_v3': 64,
+            'Standard_NC24ads_A100_v4': 80,
+            'Standard_NC48ads_A100_v4': 160,
+            'Standard_NC96ads_A100_v4': 320,
+            'Standard_ND96asr_v4': 40,
+            'Standard_ND96amsr_A100_v4': 80,
+            'Standard_ND6s': 24,
+            'Standard_ND12s': 48,
+            'Standard_ND24s': 96,
+            'Standard_ND24rs*': 96,
```
Comment (infwinston) [inline sky/clouds/service_catalog/data_fetchers/fetch_azure.py:275]: why is there a `*`?

File: sky/clouds/service_catalog/data_fetchers/fetch_azure.py [hunk 5]
```diff
@@ -244,11 +244,79 @@ def get_additional_columns(row):
         axis='columns',
     )
 
+    def create_gpu_map(df):
+        # Map of Azure's machine with GPU to their corresponding memory
+        # Result is hard-coded since Azure's API to not return such info
+        # may be outdated so need to be maintained
+        gpu_map = {
```
Comment (infwinston) [inline sky/clouds/service_catalog/data_fetchers/fetch_azure.py:251]: could we try to map instance type -> gpu name first and then calculate the resulting device memory later? this two-level approach might be cleaner.

**Review by infwinston — COMMENTED** [review]
@jc9123 thanks! the gpu memory are super useful info to the catalog. left a few comments.

File: sky/clouds/service_catalog/data_fetchers/fetch_gcp.py (hunk 1 above)
Comment (jc9123) [reply sky/clouds/service_catalog/data_fetchers/fetch_gcp.py:526]: This was a mistake on my end, I will have this removed.

File: sky/clouds/service_catalog/data_fetchers/fetch_azure.py (hunk 3 above)
Comment (jc9123) [reply sky/clouds/service_catalog/data_fetchers/fetch_azure.py:250]: I looked through the azure documentation to ensure that all instance type have been included and also ran the script on --all-regions and the result looked fine to me.
However, I think the approach you mentioned below makes more sense in which we map instance type -> gpu name  then from gpu name -> gpu memory. There is already a mapping from instance type -> gpu name in the script. Assuming this mapping is complete, we can easily map the gpu name to their corresponding memory.

File: sky/clouds/service_catalog/data_fetchers/fetch_azure.py (hunk 4 above)
Comment (jc9123) [reply sky/clouds/service_catalog/data_fetchers/fetch_azure.py:275]: Its one of the instance type offered by azure. 
https://learn.microsoft.com/en-us/azure/virtual-machines/nc-series

File: sky/clouds/service_catalog/data_fetchers/fetch_azure.py (hunk 5 above)
Comment (jc9123) [reply sky/clouds/service_catalog/data_fetchers/fetch_azure.py:251]: I agree this approach is much cleaner as it uses much less hard-coding and utilize already fetched info. I will change the script to use this approach.

**Pull request closed without merging**